File: apps/plant_shop_c-sharp/plant_shop_asp_dapper/Repositories/UserRepository.cs
using Dapper;
using plant_shop_asp_dapper.Data;
using plant_shop_asp_dapper.Models;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace plant_shop_asp_dapper.Repositories
{
    /// <summary>
    /// Repository CRUD pour les utilisateurs.
    /// </summary>
    public class UserRepository : BaseRepository
    {
        /// <summary>
        /// Constructeur avec injection de la factory.
        /// </summary>
        /// <param name="factory">Factory de connexions.</param>
        public UserRepository(DbConnectionFactory factory) : base(factory) { }

        // SQL Mappage (snake_case -> PascalCase)
        private const string SelectSql = @"
            SELECT id AS Id,
                   name AS Name,
                   email AS Email,
                   password_hash AS PasswordHash,
                   is_admin AS IsAdmin,
                   created_at AS CreatedAt
            FROM users";

        /// <summary>
        /// Trouve un utilisateur par email.
        /// </summary>
        /// <param name="email">Email a rechercher.</param>
        /// <returns>Utilisateur ou null.</returns>
        public async Task<User?> FindByEmailAsync(string email)
        {
            using var connection = CreateConnection();
            return await connection.QuerySingleOrDefaultAsync<User>(
                $"{SelectSql} WHERE email = @Email", new { Email = email });
        }

        /// <summary>
        /// Trouve un utilisateur par ID.
        /// </summary>
        /// <param name="id">ID a rechercher.</param>
        /// <returns>Utilisateur ou null.</returns>
        public async Task<User?> FindByIdAsync(int id)
        {
            using var connection = CreateConnection();
            return await connection.QuerySingleOrDefaultAsync<User>(
                $"{SelectSql} WHERE id = @Id", new { Id = id });
        }

        /// <summary>
        /// Liste tous les utilisateurs.
        /// </summary>
        /// <returns>Liste des utilisateurs.</returns>
        public async Task<IEnumerable<User>> FindAllAsync()
        {
            using var connection = CreateConnection();
            // Exclure le hash des listes
            var sql = SelectSql.Replace("password_hash AS PasswordHash,", "NULL AS PasswordHash,");
            return await connection.QueryAsync<User>(
                $"{sql} ORDER BY is_admin DESC, name ASC");
        }

        /// <summary>
        /// Cree un nouvel utilisateur.
        /// </summary>
        /// <param name="user">Utilisateur a creer.</param>
        /// <returns>Utilisateur avec ID genere.</returns>
        public async Task<User> CreateAsync(User user)
        {
            using var connection = CreateConnection();
            var sql = @"
                INSERT INTO users (name, email, password_hash, is_admin, created_at)
                VALUES (@Name, @Email, @PasswordHash, @IsAdmin, @CreatedAt)
                RETURNING id";

            user.CreatedAt = DateTime.UtcNow;
            user.Id = await connection.ExecuteScalarAsync<int>(sql, user);
            return user;
        }

        /// <summary>
        /// Met a jour un utilisateur.
        /// </summary>
        /// <param name="user">Utilisateur modifie.</param>
        public async Task UpdateAsync(User user)
        {
            using var connection = CreateConnection();
            var sql = @"
                UPDATE users
                SET name = @Name,
                    email = @Email,
                    is_admin = @IsAdmin
                WHERE id = @Id";
            await connection.ExecuteAsync(sql, user);
        }

        /// <summary>
        /// Supprime un utilisateur.
        /// </summary>
        /// <param name="id">ID a supprimer.</param>
        public async Task DeleteAsync(int id)
        {
            using var connection = CreateConnection();
            await connection.ExecuteAsync("DELETE FROM users WHERE id = @Id", new { Id = id });
        }
    }
}

File: apps/plant_shop_java/Structure_monolithique/plant_shop_java_quarkus/src/repositories/UserRepository.java
package repositories;

import jakarta.enterprise.context.Dependent;
import jakarta.inject.Inject;
import java.sql.*;
import models.User;

/**
 * Repository pour les utilisateurs.
 */
@Dependent
public class UserRepository extends BaseRepository<User> {

    /**
     * Constructeur avec injection de la connexion.
     * @param db Connexion à la base de données
     */
    @Inject
    public UserRepository(Connection db) {
        super(db, "users");
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected User mapFromResultSet(ResultSet rs) throws SQLException {
        return new User(
            rs.getInt("id"),
            rs.getString("name"),
            rs.getString("email"),
            null, // passwordHash est volontairement laissé à null
            rs.getBoolean("is_admin"),
            rs.getTimestamp("created_at")
        );
    }

    /**
     * Trouve un utilisateur par email avec le hash du mot de passe.
     * @param email Email de l'utilisateur
     * @return L'utilisateur avec passwordHash ou null
     */
    public User findByEmailWithPassword(String email) throws SQLException {
        String sql = "SELECT * FROM users WHERE email=?";
        try (PreparedStatement ps = db.prepareStatement(sql)) {
            ps.setString(1, email);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    return new User(
                        rs.getInt("id"),
                        rs.getString("name"),
                        rs.getString("email"),
                        rs.getString("password_hash"), // Le hash est inclus ici
                        rs.getBoolean("is_admin"),
                        rs.getTimestamp("created_at")
                    );
                }
                return null;
            }
        }
    }

    /**
     * Crée un nouvel utilisateur.
     * @param u Utilisateur à créer
     * @return ID de l'utilisateur créé
     */
    public int create(User u) throws SQLException {
        String sql = "INSERT INTO users(name, email, password_hash, is_admin) VALUES (?, ?, ?, ?)";
        try (PreparedStatement ps = db.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            ps.setString(1, u.name);
            ps.setString(2, u.email);
            ps.setString(3, u.passwordHash);
            ps.setBoolean(4, u.isAdmin);
            ps.executeUpdate();
            try (ResultSet rs = ps.getGeneratedKeys()) {
                rs.next();
                return rs.getInt(1);
            }
        }
    }

    /**
     * Met à jour un utilisateur existant.
     * @param u Utilisateur avec les nouvelles données
     */
    public void update(User u) throws SQLException {
        boolean updatePassword = u.passwordHash != null && !u.passwordHash.isEmpty();
        String sql = updatePassword
            ? "UPDATE users SET name=?, email=?, is_admin=?, password_hash=? WHERE id=?"
            : "UPDATE users SET name=?, email=?, is_admin=? WHERE id=?";

        try (PreparedStatement ps = db.prepareStatement(sql)) {
            ps.setString(1, u.name);
            ps.setString(2, u.email);
            ps.setBoolean(3, u.isAdmin);
            if (updatePassword) {
                ps.setString(4, u.passwordHash);
                ps.setInt(5, u.id);
            } else {
                ps.setInt(4, u.id);
            }
            ps.executeUpdate();
        }
    }
}
